lib/events/dynamoevents/dynamoevents.go

```diff
@@ -500,74 +650,87 @@ func (l *Log) SearchEvents(fromUTC, toUTC time.Time, filter string, limit int) (
-				return nil, trace.BadParameter("failed to unmarshal event for %v", err)
+		// Because the maximum size of the dynamo db response size is 900K according to documentation,
+		// we arbitrary limit the total size to 100MB to prevent runaway loops.
+	pageLoop:
+		for pageCount := 0; pageCount < 100; pageCount++ {
+			input := dynamodb.QueryInput{
+				KeyConditionExpression:    aws.String(query),
+				TableName:                 aws.String(l.Tablename),
+				ExpressionAttributeValues: attributeValues,
+				IndexName:                 aws.String(indexTimeSearchV2),
+				ExclusiveStartKey:         lastEvaluatedKey,
 			}
-			var fields events.EventFields
-			data := []byte(e.Fields)
-			if err := json.Unmarshal(data, &fields); err != nil {
-				return nil, trace.BadParameter("failed to unmarshal event %v", err)
+			start := time.Now()
+			out, err := l.svc.Query(&input)
+			if err != nil {
+				return nil, trace.Wrap(err)
 			}
-			var accepted bool
-			for i := range eventFilter {
-				if fields.GetString(events.EventType) == eventFilter[i] {
-					accepted = true
-					break
+			g.WithFields(log.Fields{"duration": time.Since(start), "items": len(out.Items)}).Debugf("Query completed.")
+
+		itemLoop:
+			for _, item := range out.Items {
+				var e event
+				if err := dynamodbattribute.UnmarshalMap(item, &e); err != nil {
+					return nil, trace.BadParameter("failed to unmarshal event for %v", err)
 				}
-			}
-			if accepted || !doFilter {
-				values = append(values, fields)
-				total++
-				if limit > 0 && total >= limit {
-					break
+				var fields events.EventFields
+				data := []byte(e.Fields)
+				if err := json.Unmarshal(data, &fields); err != nil {
+					return nil, trace.BadParameter("failed to unmarshal event %v", err)
+				}
+				var accepted bool
+				for i := range eventFilter {
+					if fields.GetString(events.EventType) == eventFilter[i] {
+						accepted = true
+						break itemLoop
+					}
+				}
+				if accepted || !doFilter {
+					values = append(values, fields)
+					total++
+					if limit > 0 && total >= limit {
+						break dateLoop
+					}
 				}
 			}
-		}
 
-		// AWS returns a `lastEvaluatedKey` in case the response is truncated, i.e. needs to be fetched with
-		// multiple requests. According to their documentation, the final response is signaled by not setting
-		// this value - therefore we use it as our break condition.
-		lastEvaluatedKey = out.LastEvaluatedKey
-		if len(lastEvaluatedKey) == 0 {
-			sort.Sort(events.ByTimeAndIndex(values))
-			return values, nil
+			// AWS returns a `lastEvaluatedKey` in case the response is truncated, i.e. needs to be fetched with
+			// multiple requests. According to their documentation, the final response is signaled by not setting
+			// this value - therefore we use it as our break condition.
+			lastEvaluatedKey = out.LastEvaluatedKey
+			if len(lastEvaluatedKey) == 0 {
+				sort.Sort(events.ByTimeAndIndex(values))
```

Comment:
Sort values once at the end instead of during every iteration of `dateLoop`.  Not only will this be more efficient, but its necessary to consistently sort the returned values since the loop may terminate early and not reach this point if `limit` was set.